Zend/zend_execute.h

```diff
@@ -167,117 +166,121 @@ ZEND_API void zend_vm_stack_init(void);
-		used_stack += func->op_array.last_var + func->op_array.T - MIN(func->op_array.num_args, num_args);
+	if (UNEXPECTED(func->type == ZEND_USER_FUNCTION && num_args < func->op_array.num_args)) {
+		num_args = func->op_array.num_args;
 	}
-	return used_stack * sizeof(zval);
-}
 
-static zend_always_inline zend_execute_data *zend_vm_stack_push_call_frame(uint32_t call_info, zend_function *func, uint32_t num_args, zend_class_entry *called_scope, zend_object *object)
-{
-	uint32_t used_stack = zend_vm_calc_used_stack(num_args, func);
+	if (EXPECTED(execute_data)) {
+		execute_data = (zend_execute_data *) (((char *) execute_data) + zend_vm_calc_used_stack(num_args, EX(func)));
+	} else {
+		execute_data = (zend_execute_data *) (ZEND_VM_STACK_ELEMENTS(EG(vm_stack)) + num_args);
+	}
 
-	return zend_vm_stack_push_call_frame_ex(used_stack, call_info,
-		func, num_args, called_scope, object);
+	return zend_push_call_frame(execute_data, call_info, func, passed_args, called_scope, object);
 }
 
-static zend_always_inline void zend_vm_stack_free_extra_args_ex(uint32_t call_info, zend_execute_data *call)
+/* internal calls *and* cleanup of not yet called user funcs */
+static zend_always_inline void zend_free_args_internal(zend_execute_data *call, uint32_t num_args)
 {
-	if (UNEXPECTED(call_info & ZEND_CALL_FREE_EXTRA_ARGS)) {
-		zval *end = ZEND_CALL_VAR_NUM(call, call->func->op_array.last_var + call->func->op_array.T);
- 		zval *p = end + (ZEND_CALL_NUM_ARGS(call) - call->func->op_array.num_args);
+	if (EXPECTED(num_args > 0)) {
+		zval *p = ZEND_CALL_ARG(call, 0);
+		zval *end = p - num_args;
+
 		do {
 			p--;
 			if (Z_REFCOUNTED_P(p)) {
 				if (!Z_DELREF_P(p)) {
 					zend_refcounted *r = Z_COUNTED_P(p);
-					ZVAL_NULL(p);
+					ZVAL_NULL(p); /* TODO: check if still necessary??? */
 					zval_dtor_func_for_ptr(r);
-				} else {
-					GC_ZVAL_CHECK_POSSIBLE_ROOT(p);
 				}
 			}
 		} while (p != end);
- 	}
-}
-
-static zend_always_inline void zend_vm_stack_free_extra_args(zend_execute_data *call)
-{
-	zend_vm_stack_free_extra_args_ex(ZEND_CALL_INFO(call), call);
+	}
 }
 
-static zend_always_inline void zend_vm_stack_free_args(zend_execute_data *call)
+static zend_always_inline void zend_free_args_user(zend_execute_data *call)
 {
-	uint32_t num_args = ZEND_CALL_NUM_ARGS(call);
+	uint32_t passed_args = ZEND_CALL_NUM_ARGS(call);
+	uint32_t num_args = call->func->op_array.num_args;
+	/* Usually we don't push more args than available, but often less (default parameter values) */
+	if (UNEXPECTED(num_args < passed_args)) {
+		num_args = passed_args;
+	}
 
 	if (EXPECTED(num_args > 0)) {
-		zval *end = ZEND_CALL_ARG(call, 1);
-		zval *p = end + num_args;
+		zval *p = ZEND_CALL_ARG(call, 0);
+		zval *end = p - num_args;
 
 		do {
 			p--;
 			if (Z_REFCOUNTED_P(p)) {
 				if (!Z_DELREF_P(p)) {
 					zend_refcounted *r = Z_COUNTED_P(p);
-					ZVAL_NULL(p);
+					ZVAL_NULL(p); /* TODO: check if still necessary??? */
 					zval_dtor_func_for_ptr(r);
```

Comment:
Indeed, we call zend_free_args_user() right before the EG(current_execute_data) assignment in leave_helper. We could move it two lines down, then it'll be fine.